src/platforms/linux/daemon/dbusservice.cpp

```diff
@@ -130,11 +136,162 @@ QString DBusService::getLogs() {
 }
 
 bool DBusService::run(Op op, const Config& config) {
+  QStringList addresses;
+  for (const IPAddressRange& ip : config.m_allowedIPAddressRanges) {
+    addresses.append(ip.toString());
+  }
+
   return WgQuickProcess::run(
       op, config.m_privateKey, config.m_deviceIpv4Address,
       config.m_deviceIpv6Address, config.m_serverIpv4Gateway,
       config.m_serverIpv6Gateway, config.m_serverPublicKey,
       config.m_serverIpv4AddrIn, config.m_serverIpv6AddrIn,
-      config.m_allowedIPAddressRanges.join(", "), config.m_serverPort,
-      config.m_ipv6Enabled);
+      addresses.join(", "), config.m_serverPort, config.m_ipv6Enabled);
+}
+
+static inline bool parseEndpoint(struct sockaddr* endpoint,
+                                 const QString& address, int port) {
+  QString portString = QString::number(port);
+
+  struct addrinfo hints;
+  memset(&hints, 0, sizeof(hints));
+  hints.ai_family = AF_UNSPEC;
+  hints.ai_socktype = SOCK_DGRAM;
+  hints.ai_protocol = IPPROTO_UDP;
+
+  struct addrinfo* resolved = nullptr;
+  int retries = 15;
+
+  for (unsigned int timeout = 1000000;;
+       timeout = std::min((unsigned int)20000000, timeout * 6 / 5)) {
+    int rv = getaddrinfo(address.toLocal8Bit(), portString.toLocal8Bit(),
+                         &hints, &resolved);
+    if (!rv) {
+      break;
+    }
+
+    /* The set of return codes that are "permanent failures". All other
+     * possibilities are potentially transient.
+     *
+     * This is according to https://sourceware.org/glibc/wiki/NameResolver which
+     * states: "From the perspective of the application that calls getaddrinfo()
+     * it perhaps doesn't matter that much since EAI_FAIL, EAI_NONAME and
+     * EAI_NODATA are all permanent failure codes and the causes are all
+     * permanent failures in the sense that there is no point in retrying
+     * later."
+     *
+     * So this is what we do, except FreeBSD removed EAI_NODATA some time ago,
+     * so that's conditional.
+     */
+    if (rv == EAI_NONAME || rv == EAI_FAIL ||
+#ifdef EAI_NODATA
+        rv == EAI_NODATA ||
+#endif
+        (retries >= 0 && !retries--)) {
+      logger.log() << "Failed to resolve the address endpoint";
+      return false;
+    }
+
+    logger.log() << "Trying again in" << (timeout / 1000000.0) << "seconds";
+    usleep(timeout);
+  }
+
+  if ((resolved->ai_family == AF_INET &&
+       resolved->ai_addrlen == sizeof(struct sockaddr_in)) ||
+      (resolved->ai_family == AF_INET6 &&
+       resolved->ai_addrlen == sizeof(struct sockaddr_in6))) {
+    memcpy(endpoint, resolved->ai_addr, resolved->ai_addrlen);
+    freeaddrinfo(resolved);
+    return true;
+  }
+
+  logger.log() << "Invalid endpoint" << address;
+  freeaddrinfo(resolved);
+  return false;
```

Comment:
A style question: The ending of this method stood out because it duplicates the call to freeaddrinfo and we end on the unhappy path....as opposed to maybe returning a "success" variable....although I can see why it is what it is. 

There's a lot of methods like this, that change states but return true/false. Maybe it's a little bit of a functional programming hangover, or a wariness of "side-effects."

I'm just throwing these thoughts out there are discussion points in case you have input / thoughts.

Reply:
Yes, I see your point. We could expose a sockaddr struct or nullptr. But then we need to copy its content into the peer->endpoint.addr...